WalletWasabi/Extensions/StreamExtensions.cs

```diff
@@ -1,32 +1,66 @@
+using System.Buffers;
 using System.Threading;
 using System.Threading.Tasks;
-using System;
 
 namespace System.IO
 {
 	public static class StreamExtensions
 	{
 		public static async Task<int> ReadByteAsync(this Stream stream, CancellationToken ctsToken = default)
 		{
-			var buf = new byte[1];
-			var len = await stream.ReadAsync(buf.AsMemory(0, 1), ctsToken).ConfigureAwait(false);
-			if (len == 0)
+			ArrayPool<byte> pool = ArrayPool<byte>.Shared;
+			byte[] buffer = pool.Rent(1);
+			try
 			{
-				return -1;
-			}
+				int len = await stream.ReadAsync(buffer.AsMemory(0, 1), ctsToken).ConfigureAwait(false);
+
+				// End of stream.
+				if (len == 0)
+				{
+					return -1;
+				}
 
-			return buf[0];
+				return buffer[0];
+			}
+			finally
+			{
+				pool.Return(buffer);
+			}
 		}
 
 		public static async Task<int> ReadBlockAsync(this Stream stream, byte[] buffer, int count, CancellationToken ctsToken = default)
 		{
-			var left = count;
+			int left = count;
 			while (left != 0)
 			{
-				var read = await stream.ReadAsync(buffer.AsMemory(count - left, left), ctsToken).ConfigureAwait(false);
+				int read = await stream.ReadAsync(buffer.AsMemory(count - left, left), ctsToken).ConfigureAwait(false);
 				left -= read;
 			}
 			return count - left;
 		}
+
+		/// <summary>
+		/// Reads exactly <paramref name="count"/> bytes from <paramref name="stream"/>.
+		/// </summary>
+		/// <param name="stream">Stream to read from.</param>
+		/// <param name="buffer">Buffer whose length must be at least <paramref name="count"/> elements.</param>
+		/// <param name="count">Number of bytes to read.</param>
+		/// <param name="cancellationToken">Cancellation token to cancel the asynchronous operation.</param>
+		/// <returns><c>true</c> if we could read exactly <paramref name="count"/> bytes from stream (stream may contain more bytes though).</returns>
+		public static async Task<bool> ReadExactlyAsync(this Stream stream, byte[] buffer, int count, CancellationToken cancellationToken = default)
```

Comment:
This method is conceptually incorrect. A stream cannot be read "exactly" because you can never know how the sender is writing.